rixs/process2d.py

```diff
@@ -0,0 +1,240 @@
+              (y_edges=x_edges=bins).
+            * If [int, int], the number of bins in each dimension
+              (ny, nx = bins).
+            * If [array, array], the bin edges in each dimension
+              (y_edges, x_edges = bins).
+            * A combination [int, array] or [array, int], where int
+              is the number of bins and array is the bin edges.
+
+    Returns
+    -------------
+    x_centers : array
+        columns positions where offsets were determined
+        i.e. binx/2, 3*binx/2, 5*binx/2, ...
+    offsets : array
+        np.array of row offsets defining curvature. This is referenced
+        to the center of the image.
+    """
+    x = photon_events[:, 0]
+    y = photon_events[:, 1]
+    Iph = photon_events[:, 2]
+
+    if bins is None:
+        ybins = np.arange(y.min()//1 + 0.5, y.max()//1, 1)
+        xstep = 50
+        xbins = np.arange(x.min() + xstep/2, x.max() - xstep/2, xstep)
+        bins = (ybins, xbins)
+
+    image, y_edges, x_edges = np.histogram2d(y, x, bins=bins, weights=Iph)
+    y_centers = (y_edges[:-1] + y_edges[1:])/2
+    x_centers = (x_edges[:-1] + x_edges[1:])/2
+
+    ref_column = image[:, image.shape[1]//2]
+
+    offsets = np.array([])
+    for column in image.T:
+        cross_correlation = np.correlate(column, ref_column, mode='same')
+        offsets = np.append(offsets, y_centers[np.argmax(cross_correlation)])
+    return x_centers, offsets - offsets[offsets.shape[0]//2]
+
+
+def fit_curvature(photon_events, guess_curvature, bins=None):
+    """Get offsets, fit them and return polynomial that defines the curvature
+
+    Parameters
+    -------------
+    photon_events : array
+        two column x, y photon location coordinates
+    curvature : array
+        The polynominal coeffcients describing the image curvature.
+        These are in decreasing order e.g.
+        curvature[0]*x^2 + curvature[1]*x^1, + curvature[2]*x^0, `
+    bins : int or array_like or [int, int] or [array, array]
+        The bin specification in y then x order:
+            * If int, the number of bins for the two dimensions (nx=ny=bins).
+            * If array_like, the bin edges for the two dimensions
+              (y_edges=x_edges=bins).
+            * If [int, int], the number of bins in each dimension
+              (ny, nx = bins).
+            * If [array, array], the bin edges in each dimension
+              (y_edges, x_edges = bins).
+            * A combination [int, array] or [array, int], where int
+              is the number of bins and array is the bin edges.
+
+    Returns
+    -----------
+
+    """
+    x_centers, offsets = get_curvature_offsets(photon_events, bins)
+    reversed_curvature = polyfit(x_centers, offsets, len(guess_curvature) - 1)
+    curvature = reversed_curvature[::-1]
+
+    return curvature
+
+
+def estimate_elastic_pos(photon_events, x_range=(0, 20), bins=None):
+    """
+    Estimate where the elastic line is.
+
```

Comment:
I think we should remove one of these empty lines

Reply:
sure